replica-node/src/replica_util.rs

```diff
@@ -0,0 +1,268 @@
+use {
+    log::*,
+    rand::{seq::SliceRandom, thread_rng, Rng},
+    solana_gossip::{
+        cluster_info::{ClusterInfo, Node},
+        contact_info::ContactInfo,
+        gossip_service::GossipService,
+    },
+    solana_runtime::snapshot_utils,
+    solana_sdk::{
+        clock::Slot,
+        hash::Hash,
+        pubkey::Pubkey,
+        signature::{Keypair, Signer},
+    },
+    std::{
+        collections::HashSet,
+        net::{SocketAddr, UdpSocket},
+        path::Path,
+        process::exit,
+        sync::{atomic::AtomicBool, Arc},
+        thread::sleep,
+        time::{Duration, Instant},
+    },
+};
+
+pub fn get_cluster_shred_version(entrypoints: &[SocketAddr]) -> Option<u16> {
+    let entrypoints = {
+        let mut index: Vec<_> = (0..entrypoints.len()).collect();
+        index.shuffle(&mut rand::thread_rng());
+        index.into_iter().map(|i| &entrypoints[i])
+    };
+    for entrypoint in entrypoints {
+        match solana_net_utils::get_cluster_shred_version(entrypoint) {
+            Err(err) => eprintln!("get_cluster_shred_version failed: {}, {}", entrypoint, err),
+            Ok(0) => eprintln!("zero sherd-version from entrypoint: {}", entrypoint),
+            Ok(shred_version) => {
+                info!(
+                    "obtained shred-version {} from {}",
+                    shred_version, entrypoint
+                );
+                return Some(shred_version);
+            }
+        }
+    }
+    None
+}
+
+// Discover the RPC peer node via Gossip and return's ContactInfo
+// And the initial snapshot info: (Slot, Hash)
+// Alternatively, this can be solved via a RPC call instead of using gossip.
+fn get_rpc_peer_node(
```

Comment:
Can we dedupe any of this with the existing `get_rpc_node()`? I haven't diffed them to see how different they are, but at a glance there looks like a fair bit in common.

Reply:
This code is adapted from get_rpc_node. For the same reason as above -- I would delay refactoring this with validator. The code will be removed most likely in another PR when snapshot can be discovered via RPC instead of gossip.